python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -211,13 +211,17 @@ def _update(self):
         if self.resource_demand_scheduler:
             resource_demand_vector = self.resource_demand_vector + \
                 self.load_metrics.get_resource_demand_vector()
+            non_terminated_nodes = self.provider.non_terminated_nodes(
+                tag_filters={})
+            pending_launches_nodes = self.pending_launches.breakdown()
             to_launch = (self.resource_demand_scheduler.get_nodes_to_launch(
-                self.provider.non_terminated_nodes(tag_filters={}),
-                self.pending_launches.breakdown(),
+                non_terminated_nodes, pending_launches_nodes,
                 resource_demand_vector,
                 self.load_metrics.get_resource_utilization()))
-            # TODO(ekl) also enforce max launch concurrency here?
-            for node_type, count in to_launch.items():
+
+            updated_to_launch = self._get_concurrent_resource_demand_to_launch(
+                to_launch, non_terminated_nodes, pending_launches_nodes)
+            for node_type, count in updated_to_launch.items():
```

Comment:
Can you actually move this logic into resource_demand_scheduler instead? That way we can unit test it easily in test_resource_demand_scheduler independent of the big autoscaler class.

E.g., as in test_get_nodes_to_launch_with_min_workers